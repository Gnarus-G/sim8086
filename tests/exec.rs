use std::process::Command;

use insta::assert_display_snapshot;

macro_rules! test_with {
    ($file:literal) => {
        let _ = Command::new("nasm")
            .arg(format!("./fixtures/exec/{}.asm", $file))
            .status()
            .unwrap();

        let app_output = assert_cmd::Command::cargo_bin("sim8086")
            .unwrap()
            .arg(format!("./fixtures/exec/{}", $file))
            .arg("--exec")
            .output()
            .map(|out| String::from_utf8(out.stdout).unwrap())
            .unwrap();

        insta::with_settings!({ description => $file }, {
            assert_display_snapshot!(app_output);
        })
    };
}

#[test]
fn movs() {
    test_with!("immediate_movs");
    test_with!("register_movs");
}

#[test]
fn add_sub_cmp() {
    test_with!("add_sub_cmp");
}

#[test]
fn ip_tracking() {
    test_with!("ip_register");
}

#[test]
fn conditional_jumps() {
    // test_with!("add_sub_cmp");
}
