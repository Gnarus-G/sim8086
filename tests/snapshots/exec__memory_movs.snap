---
source: tests/exec.rs
description: memory_mov
expression: app_output
---
mov [1000], word 1 ; ip:0x0->0x6, 
mov [1002], word 2 ; ip:0x6->0xc, 
mov [1004], word 3 ; ip:0xc->0x12, 
mov [1006], word 4 ; ip:0x12->0x18, 
mov bx, 1000 ; bx:0x0->0x3e8, ip:0x18->0x1b, 
mov [bx + 4], word 10 ; ip:0x1b->0x20, 
mov bx, [1000] ; bx:0x3e8->0x1, ip:0x20->0x24, 
mov cx, [1002] ; cx:0x0->0x2, ip:0x24->0x28, 
mov dx, [1004] ; dx:0x0->0xa, ip:0x28->0x2c, 
mov bp, [1006] ; bp:0x0->0x4, ip:0x2c->0x30, 

Final registers:
   ax: 0x0000 (0)
   bx: 0x0001 (1)
   cx: 0x0002 (2)
   dx: 0x000a (10)
   sp: 0x0000 (0)
   bp: 0x0004 (4)
   si: 0x0000 (0)
   di: 0x0000 (0)
   ip: 0x0030 (48)
flags: 

